Celeste.Mod.mm/Mod/UI/OuiModTogglerProgress.cs

```diff
@@ -0,0 +1,81 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Celeste.Mod.UI {
+    public class OuiModTogglerProgress : OuiLoggedProgress {
+
+        private List<string> newlyUnblacklistedMods;
+
+        public void Init(List<string> mods) {
+            Everest.Loader.OnCrawlMod += logCrawlMod;
+            newlyUnblacklistedMods = mods;
+            Init<OuiMainMenu>(Dialog.Clean("MODOPTIONS_MODTOGGLE_PROGRESS_TITLE"), new Task(toggleMods),
+                newlyUnblacklistedMods.Count);
+        }
+
+        public override IEnumerator Leave(Oui next) {
+            Everest.Loader.OnCrawlMod -= logCrawlMod;
+            newlyUnblacklistedMods = null;
+            MainThreadHelper.Do(() => ((patch_OuiMainMenu) Overworld.GetUI<OuiMainMenu>())?.RebuildMainAndTitle());
+            Audio.Play(SFX.ui_main_button_back);
+            
+            return base.Leave(next);
+        }
+
+        private void toggleMods() {
+            if (newlyUnblacklistedMods == null)
+                return;
+            
+            // give it a second to transition
+            Thread.Sleep(1000);
```

Comment:
Does this serve a functional purpose or is it just aesthetic? if it is required then could we instead wait to initialize the Task until the end of the `Enter` coroutine?

Reply:
its pretty much required since the whole UI locks up before the transition finishes--but yes that seems like a better solution

Reply:
okay i think doing this will actually turn out to be way uglier than just sleeping

Reply:
What about something like this? OuiLoggedProgress won't call `OnFinished` until the Task is no longer null *and* is completed.
```diff
diff --git a/Celeste.Mod.mm/Mod/UI/OuiLoggedProgress.cs b/Celeste.Mod.mm/Mod/UI/OuiLoggedProgress.cs
index dc215a89..a2ee34f0 100644
--- a/Celeste.Mod.mm/Mod/UI/OuiLoggedProgress.cs
+++ b/Celeste.Mod.mm/Mod/UI/OuiLoggedProgress.cs
@@ -44,7 +44,7 @@ namespace Celeste.Mod.UI {
 
             OnFinish += () => Overworld.Goto<T>();
 
-            if (task.Status == TaskStatus.Created)
+            if (task?.Status == TaskStatus.Created)
                 task.Start();
 
             return this;
diff --git a/Celeste.Mod.mm/Mod/UI/OuiModTogglerProgress.cs b/Celeste.Mod.mm/Mod/UI/OuiModTogglerProgress.cs
index f4eadf76..571323bd 100644
--- a/Celeste.Mod.mm/Mod/UI/OuiModTogglerProgress.cs
+++ b/Celeste.Mod.mm/Mod/UI/OuiModTogglerProgress.cs
@@ -14,10 +14,17 @@ namespace Celeste.Mod.UI {
         public void Init(List<string> mods) {
             Everest.Loader.OnCrawlMod += logCrawlMod;
             newlyUnblacklistedMods = mods;
-            Init<OuiMainMenu>(Dialog.Clean("MODOPTIONS_MODTOGGLE_PROGRESS_TITLE"), new Task(toggleMods),
+            Init<OuiMainMenu>(Dialog.Clean("MODOPTIONS_MODTOGGLE_PROGRESS_TITLE"), default(Task),
                 newlyUnblacklistedMods.Count);
         }
 
+        public override IEnumerator Enter(Oui prev) {
+            yield return base.Enter(prev);
+            // Delay starting until the transition is finished
+            Task = new Task(toggleMods, 0);
+            Task.Start();
+        }
+
         public override IEnumerator Leave(Oui next) {
             Everest.Loader.OnCrawlMod -= logCrawlMod;
             newlyUnblacklistedMods = null;
@@ -30,9 +37,7 @@ namespace Celeste.Mod.UI {
         private void toggleMods() {
             if (newlyUnblacklistedMods == null)
                 return;
-            
-            // give it a second to transition
-            Thread.Sleep(1000);
+
             int oldDelayedMods = Everest.Loader.Delayed.Count;
             Everest.Loader.EnforceOptionalDependencies = true;
             foreach (string mod in newlyUnblacklistedMods) {
```

Reply:
Did you ever end up looking into this?

Reply:
i did and could not get it to work